refactor(VRipple): Use PointerEvent to fix some bugs

Jesse205 opened the pull request:
<!--
MAKE SURE TO READ THE CONTRIBUTING GUIDE BEFORE CREATING A PR
https://vuetifyjs.com/getting-started/contributing

Provide a general summary of your changes in the title above
Keep the title short and descriptive, as it will be used as a commit message
PR titles should follow conventional-changelog-angular:
https://vuetifyjs.com/getting-started/contributing/#commit-guidelines
-->

## Description
<!--
Describe your changes in detail. Why is this change required? What problem does it solve?
If it fixes an open issue, please link to the issue here.
e.g. resolves #4213 or fixes #2312
-->

~~fixes #21468~~

In newer browsers, `PointerEvent` can be used instead of `MouseEvent` and `TouchEvent`.

Compatibility list: https://caniuse.com/?search=PointerEvent

---

### Changes

- [X] fixes #21468
- [X] I also fixed a bug that existed in previous versions where pressing and holding two keys simultaneously and then releasing them would cause the ripple effect not to disappear.
- [X] I also made the ripple effect caused by pressing multiple keys into a single ripple.
- [X] I replaced `KeyboardEvent.keyCode` with the updated `KeyboardEvent.key` and fixed an error in `keyValues`.
- [X] I also took the opportunity to fix a bug where the ripple effect would still appear when scrolling the page, making it consistent with Android's behavior.

### New Bugs

- [ ] https://github.com/vuetifyjs/vuetify/pull/21469#issuecomment-3129883325

## Markup:
<!--
Information on how to set up your local development environment can be found here:
https://vuetifyjs.com/getting-started/contributing/#setting-up-your-environment
Remove this section for documentation or test-only changes.
-->

<!-- Paste your FULL packages/vuetify/dev/Playground.vue here --->
```vue
<template>
  <v-container
    :style="{
      touchAction: isScrollable ? 'manipulation' : 'none',
    }"
  >
    <div
      :style="{
        position: 'sticky',
        top: 0,
        backgroundColor: current.colors.background,
        zIndex: 1,
      }"
    >
      <v-switch v-model="isScrollable" label="Scrollable" color="primary" />
    </div>

    <v-row>
      <v-col cols="12">
        <v-card v-ripple @click="count++" style="user-select: none;">
          <v-card-text>Count: {{ count }}</v-card-text>
          <v-card-actions>
            <v-btn @click.stop>Button with ripple</v-btn>
          </v-card-actions>
        </v-card>
      </v-col>
      <v-col cols="12">
        <v-card v-ripple>
          <v-card-text>
            Outer card with ripple
            <v-card-actions>
              <v-card v-ripple.stop>
                <v-card-text>
                  Inner card with <code>ripple.stop</code>
                </v-card-text>
              </v-card>
            </v-card-actions>
          </v-card-text>
        </v-card>
      </v-col>
    </v-row>

    <div>
      <v-btn
        v-ripple
        v-for="item in 500"
        text="Edit Profile"
        variant="text"
        style="border: none; border-radius: initial"
      />
    </div>
  </v-container>
</template>

<script setup>
import { useTheme } from "@/composables";
import { ref } from "vue";
const isScrollable = ref(true);
const count = ref(0);
const { current } = useTheme();
</script>


```

Jesse205 committed:
fix(VRipple): allow mouse triggering after touch triggering.

Solve this problem using PointerEvent.

fixes #21468

KaelWD commented:
We don't support anything without PointerEvent anyway: https://vuetifyjs.com/en/getting-started/browser-support/
The ripple code is all ancient, I think the last big change was in like 2018.

Jesse205 committed:
fix: remove older support and fix bugs

Jesse205 committed:
fix: keyValues space error

Jesse205 committed:
feat: no delay using mouse

Jesse205 committed:
refactor: remove console log

Jesse205 committed:
fix: ripple hide immediately with touch

Jesse205 committed:
fix: touch inner and outer at one time

Jesse205 commented:
- I have removed the code related to mouse events and touch events.
- I also fixed a bug that existed in previous versions where pressing and holding two keys simultaneously and then releasing them would cause the ripple effect not to disappear.
- I also made the ripple effect caused by pressing multiple keys into a single ripple.
- I replaced `KeyboardEvent.keyCode` with the updated `KeyboardEvent.key` and fixed an error in `keyValues`.
- I also took the opportunity to fix a bug where the ripple effect would still appear when scrolling the page, making it consistent with Android's behavior.

---

I am not familiar with Vitest, and I don't know how to fix the tests. I hope you can help me.

KaelWD committed:
style: fix lint

KaelWD committed:
use a Set

KaelWD committed:
fix includes call

KaelWD committed:
Merge branch 'master' into fix/ripple-touch-mouse

J-Sek committed:
Merge branch 'master' into fix/ripple-touch-mouse

J-Sek commented:
It would need to be verified on an actual device, but when simulating the pinch-zoom (holding `shift` when moving the pointer in Chrome DevTools mobile emulation)... there is a minor regression because ripple is triggered (previously it was not).

J-Sek committed:
add `keyUpHandler` to `_ripple`

J-Sek committed:
drop misc type changes

J-Sek committed:
drop misc type changes

Jesse205 commented:
> It would need to be verified on an actual device, but when simulating the pinch-zoom (holding `shift` when moving the pointer in Chrome DevTools mobile emulation)... there is a minor regression because ripple is triggered (previously it was not).

@J-Sek I believe this is a bug in Chromium. I tested Edge, Chrome, and Cent Browser on Windows. When simulating a device, holding down the <kbd>shift</kbd> key and moving the mouse does not trigger `pointercancel`; instead, it continues to respond to `pointermove` events.

Fortunately, Chromium on Android does not have this issue. I tested Edge, Chrome, and Cromite on Android, and when scaling begins, they stop responding to `pointermove` and correctly trigger `pointercancel`. Therefore, as long as the time between pressing and starting scaling is short enough (less than `DELAY_RIPPLE`), the ripple effect will not appear.

Firefox on Android also does not have this issue, as it responds with a `pointermove` followed by `pointercancel` when two fingers are pressed, so there is no problem in this scenario either.

J-Sek commented:
Thanks for putting the time for manual verification. Once E2E tests are green I will double-check and we should be ready for merge.
